rep-2011.rst

```diff
@@ -321,7 +321,214 @@ The passed ``rosidl_message_type_support_t`` in the init call can be introspecte
+Serialization library support in this case will be provided by writing plugins that wrap the serialization libraries that can then plug in dynamically to the runtime type introspection feature using a unified plugin interface. Then, runtime type introspection will:
+
+- enumerate the supported serialization library plugins on the machine
+- match, if possible, the serialization format specified in the message type description to an appropriate plugin (throwing a warning otherwise)
+- dynamically load the plugin
+
+These serialization library plugins must wrap functionality to:
+
+- determine suitability for use with a message's reported serialization version and type
+- parse the ``.idl`` / ``.msg`` component of the message description into a serialization library usable form
+- given that parsed description (which provides type reflection information about the message buffer), deserialize the message buffer using the serialization library
+
+In particular, the step of parsing the ``.idl``  / ``.msg`` file is necessary because there might be serialization library or type support specific steps or considerations (e.g. field re-ordering, endianness) that would not be captured in the ``.idl`` / ``.msg`` file, because they would have been assumed to have been applied during type support.
+
+Dealing with Multiple Applicable Plugins for A Serialization Format
+"""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
+
+In the case where there exists multiple applicable plugins for a particular serialization format, the plugin matching should follow this priority order:
+
+- user specified overrides passed to the matching function
+- defaults defined in the plugin matching function, if applicable, otherwise
+- the first suitable plugin in alphanumeric sorting order
+
+
+
+Example Plugin Interface
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Plugin Matching and Loading
+"""""""""""""""""""""""""""
+
+The following is an example of how this plugin matching and loading interface could look like, defining new ``rcl`` interfaces; with a plugin wrapping FastCDR v1.0.24 for serialization, with some arbitrary message type `Foo`:
+
+.. code::
+
+    // Suppose FooDescription reports that it uses FastCDR v1.0.24 for its serialization
+    rcl_message_description_t FooDescription = node->get_type_description("/foo_endpoint_name");
+
+    rcl_type_introspection_t * introspection_handle;
+    introspection_handle->init();  // Locate local plugins here
+
+    // Plugin name: "fastcdr_v1_0_24"
+    const char * plugin_name = introspection_handle->match_plugin(FooDescription->get_serialization_format());
+    rcl_serialization_plugin_t * plugin = introspection_handle->load_plugin("fastcdr_v1_0_24");
+
+    // If we wanted to force the use of MicroCDR instead
+    introspection_handle->match_plugin(FooDescription->get_serialization_type(), "microcdr");
```

Comment:
Let's try using a concrete example, like `/scan` and `sensor_msgs/msg/LaserScan`. (or image or point cloud)

Reply:
https://github.com/wjwwood/rep/pull/3/commits/bdfa2cfcd926334b1f28845d91b66867ab9218a6